### PR TITLE
Device side tests

### PR DESCRIPTION
**Pull request by neon60**
Implemented multiple device side tests and updated the workarounds.
There are three failures and those tests are commented out:
-TestStablePartitionDevice 
-TestSortByKeyDevice 
-TestStableSortByKeyDevice

### CODE REVIEW

File: test/test_partition.cpp [hunk 1]
```diff
@@ -1796,3 +1796,158 @@ TEST(PartitionTests, TestStablePartitionCopyStencilDispatchImplicit)
...
+              ASSERT_EQ(h_true, d_true);
+              ASSERT_EQ(h_false, d_false);
+          }
+      }
+}
+
+//TODO: Tests fails with error message "Memory access fault by GPU node-1 (Agent handle: 0x1ade7d0) on address 0x7fac25a00000. Reason: Page not present or supervisor privilege."
+
+// __global__
+// THRUST_HIP_LAUNCH_BOUNDS_DEFAULT
```
Comment (saadrahim) [inline test/test_partition.cpp:1912]: Is this a bug that needs to be filed?

File: test/test_for_each.cpp [hunk 2]
```diff
@@ -96,7 +96,7 @@ THRUST_HIP_LAUNCH_BOUNDS_DEFAULT
 void simple_test_kernel(F func, int size)
 {
     // (void) func; (void) size;
-    thrust::for_each(thrust::seq,
+    thrust::for_each(thrust::hip::par,
```
Comment (saadrahim) [inline test/test_for_each.cpp:99]: What is the purpose of this change?

File: test/test_sort_by_key.cpp [hunk 3]
```diff
@@ -255,56 +255,58 @@ TEST(SortByKeyTests, TestSortByKeyBoolDescending)
...
-                               thrust::raw_pointer_cast(&d_values[0]));
-
-            ASSERT_EQ(h_keys, d_keys);
-            // Only keys are compared here, the sequential stable_merge_sort that's used in
-            // CUDA and HIP don't generate the correct value sorting
-        }
-    }
-}
+//TODO: Tests fails with error message "Memory access fault by GPU node-1 (Agent handle: 0x1ade7d0) on address 0x7fac25a00000. Reason: Page not present or supervisor privilege."
+
```
Comment (saadrahim) [inline test/test_sort_by_key.cpp:259]: Do we need bug for this or is this the device malloc issue?

**Review by saadrahim — COMMENTED** [review]
Requesting some clarification.

File: test/test_partition.cpp (hunk 1 above)
Comment (neon60) [reply test/test_partition.cpp:1912]: We should check them maybe we can find something in the sequential function call. We will have to file them, probably device malloc isse.

File: test/test_for_each.cpp (hunk 2 above)
Comment (neon60) [reply test/test_for_each.cpp:99]: When we call ```thrust::hip::par``` on device side it call the thrust::seq algorithm. So with this we increased the test coverage, because we check the device side handling of the ```thrust::hip::par``` case too.

File: test/test_sort_by_key.cpp (hunk 3 above)
Comment (neon60) [reply test/test_sort_by_key.cpp:259]: We should check them in a separate issue, maybe we can find something in the sequential function call. We will have to file them, probably device malloc issue.

File: test/test_for_each.cpp (hunk 2 above)
Comment (saadrahim) [reply test/test_for_each.cpp:99]: But we removed thrust:seq as a test case. Are we ok to drop that test?

File: test/test_partition.cpp (hunk 1 above)
Comment (saadrahim) [reply test/test_partition.cpp:1912]: Please make a JIRA issue, so we do not forget about this.

File: test/test_sort_by_key.cpp (hunk 3 above)
Comment (saadrahim) [reply test/test_sort_by_key.cpp:259]: Probably should have a JIRA issue, maybe the same as the comment above.

Comment (neon60) [reply test/test_sort_by_key.cpp:259]: Wrote out the issue.

File: test/test_for_each.cpp (hunk 2 above)
Comment (neon60) [reply test/test_for_each.cpp:99]: Wrote out the issue.

File: test/test_partition.cpp (hunk 1 above)
Comment (neon60) [reply test/test_partition.cpp:1912]: Wrote out the issue.
